File: src/main/java/pl/miloszgilga/audioplayer/SchedulerActions.java
/*
 * Copyright (c) 2023 by MILOSZ GILGA <http://miloszgilga.pl>
 *
 * File name: SchedulerActions.java
 * Last modified: 17/05/2023, 01:34
 * Project name: jwizard-discord-bot
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     <http://www.apache.org/license/LICENSE-2.0>
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the license.
 */

package pl.miloszgilga.audioplayer;

import lombok.Getter;
import lombok.AccessLevel;
import lombok.extern.slf4j.Slf4j;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;

import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageEmbed;

import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.ScheduledFuture;

import pl.miloszgilga.misc.JDALog;
import pl.miloszgilga.misc.Utilities;
import pl.miloszgilga.locale.ResLocaleSet;
import pl.miloszgilga.dto.TrackPosition;
import pl.miloszgilga.dto.CommandEventWrapper;
import pl.miloszgilga.embed.EmbedMessageBuilder;
import pl.miloszgilga.core.remote.RemoteProperty;
import pl.miloszgilga.core.remote.RemotePropertyHandler;
import pl.miloszgilga.core.configuration.BotConfiguration;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@Slf4j
public class SchedulerActions {

    private final AudioPlayer audioPlayer;
    private final BotConfiguration config;
    private final EmbedMessageBuilder builder;
    private final CommandEventWrapper deliveryEvent;
    private final RemotePropertyHandler handler;

    @Getter(value = AccessLevel.PACKAGE)    private final Queue<AudioQueueExtendedInfo> trackQueue = new LinkedList<>();

    @Getter(value = AccessLevel.PUBLIC)     private AudioTrack pausedTrack;
    @Getter(value = AccessLevel.PACKAGE)    private boolean onClearing = false;
    @Getter(value = AccessLevel.PACKAGE)    private boolean infiniteRepeating = false;
    @Getter(value = AccessLevel.PACKAGE)    private boolean infinitePlaylistRepeating = false;
    @Getter(value = AccessLevel.PACKAGE)    private int countOfRepeats = 0;
    @Getter(value = AccessLevel.PACKAGE)    private boolean nextTrackInfoDisabled = false;
    @Getter(value = AccessLevel.PACKAGE)    private ScheduledFuture<?> threadCountToLeave;

    private int totalCountOfRepeats = 0;

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    SchedulerActions(
        TrackScheduler trackScheduler, BotConfiguration config, EmbedMessageBuilder builder, RemotePropertyHandler handler
    ) {
        this.audioPlayer = trackScheduler.getAudioPlayer();
        this.deliveryEvent = trackScheduler.getDeliveryEvent();
        this.config = config;
        this.builder = builder;
        this.handler = handler;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    void addToQueueAndOffer(AudioQueueExtendedInfo extendedInfo) {
        if (audioPlayer.startTrack(extendedInfo.audioTrack(), true)) return;
        trackQueue.offer(extendedInfo);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public void nextTrack() {
        final AudioQueueExtendedInfo extendedInfo = trackQueue.poll();
        if (Objects.isNull(extendedInfo)) return;
        audioPlayer.startTrack(extendedInfo.audioTrack(), false);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public void skipToPosition(int position) {
        AudioQueueExtendedInfo extendedInfo = null;
        for (int i = 0; i < position; i++) {
            extendedInfo = trackQueue.poll();
        }
        if (Objects.isNull(extendedInfo)) return;
        audioPlayer.startTrack(extendedInfo.audioTrack(), false);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public AudioTrack moveToPosition(TrackPosition positions) {
        final List<AudioQueueExtendedInfo> copyTracks = new ArrayList<>(trackQueue);
        final AudioQueueExtendedInfo selectedTrack = copyTracks.remove(positions.previous() - 1);
        copyTracks.add(positions.selected() - 1, selectedTrack);
        trackQueue.clear();
        trackQueue.addAll(copyTracks);
        return selectedTrack.audioTrack();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public List<ExtendedAudioTrackInfo> removeAllTracksFromMember(Member member) {
        final List<ExtendedAudioTrackInfo> removedTracks = new ArrayList<>();
        final List<AudioQueueExtendedInfo> toRemove = new ArrayList<>(trackQueue);
        for (final AudioQueueExtendedInfo track : toRemove) {
            if (!track.sender().equals(member)) continue;

            final AudioQueueExtendedInfo extendedInfo = trackQueue.poll();
            if (Objects.isNull(extendedInfo)) continue;
            removedTracks.add(new ExtendedAudioTrackInfo(extendedInfo.audioTrack()));
        }
        return removedTracks;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public boolean checkIfAllTrackOrTracksIsFromSelectedMember(Member member) {
        if (trackQueue.isEmpty()) {
            if (Objects.isNull(audioPlayer.getPlayingTrack())) return false;
            final Member sender = (Member) audioPlayer.getPlayingTrack().getUserData();
            return sender.getId().equals(member.getId());
        }
        return trackQueue.stream().allMatch(t -> t.sender().getId().equals(member.getId()));
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public void clearAndDestroy(boolean showMessage) {
        onClearing = true;
        audioPlayer.setPaused(false);
        audioPlayer.stopTrack();
        trackQueue.clear();

        pausedTrack = null;
        countOfRepeats = 0;
        totalCountOfRepeats = 0;
        infiniteRepeating = false;
        nextTrackInfoDisabled = false;
        infinitePlaylistRepeating = false;

        if (showMessage) {
            final MessageEmbed messageEmbed = builder.createMessage(ResLocaleSet.LEAVE_EMPTY_CHANNEL_MESS,
                deliveryEvent.getGuild());
            deliveryEvent.getTextChannel().sendMessageEmbeds(messageEmbed).queue();
        }
        onClearing = false;
        JDALog.info(log, deliveryEvent, "Remove playing track and clear queue");
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    void closeAudioConnection() {
        final Guild guild = deliveryEvent.getDataSender().getGuild();
        config.getThreadPool().submit(() -> guild.getAudioManager().closeAudioConnection());
        JDALog.info(log, deliveryEvent, "Audio connection threadpool was closed");
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public void leaveAndSendMessageAfterInactivity() {
        final int timeToLeaveChannel = handler.getPossibleRemoteProperty(RemoteProperty.R_INACTIVITY_NO_TRACK_TIMEOUT,
            deliveryEvent.getGuild(), Integer.class);
        threadCountToLeave = config.getThreadPool().schedule(() -> {
            final MessageEmbed leaveMessageEmbed = builder
                .createMessage(ResLocaleSet.LEAVE_END_PLAYBACK_QUEUE_MESS, Map.of(
                    "elapsed", Utilities.convertSecondsToMinutes(timeToLeaveChannel)
                ), deliveryEvent.getGuild());
            clearAndDestroy(false);
            closeAudioConnection();

            deliveryEvent.getTextChannel().sendMessageEmbeds(leaveMessageEmbed).queue();
            JDALog.info(log, deliveryEvent, "Leave voice channel after '%s' seconds of inactivity", timeToLeaveChannel);
        }, timeToLeaveChannel, TimeUnit.SECONDS);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    String getTrackPositionInQueue() {
        if (trackQueue.size() == 1) return config.getLocaleText(ResLocaleSet.NEXT_TRACK_INDEX_MESS, deliveryEvent.getGuild());
        return Integer.toString(trackQueue.size());
    }

    public AudioTrack getTrackByPosition(int position) {
        final List<AudioQueueExtendedInfo> audioTracks = new ArrayList<>(trackQueue);
        return audioTracks.get(position - 1).audioTrack();
    }

    public long getAverageTrackDuration() {
        return (long) (trackQueue.stream().mapToLong(t -> t.audioTrack().getDuration()).average().orElse(0));
    }

    public boolean checkInvTrackPosition(int position) {
        return position <= 0 || position > trackQueue.size();
    }

    boolean checkIfMemberAddAnyTracksToQueue(Member member) {
        return trackQueue.stream().anyMatch(t -> {
            final Member iterateMember = (Member) t.audioTrack().getUserData();
            return member.equals(iterateMember);
        });
    }

    boolean toggleInfinitePlaylistRepeating() {
        infinitePlaylistRepeating = !infinitePlaylistRepeating;
        return infinitePlaylistRepeating;
    }

    void setCountOfRepeats(int countOfRepeats) {
        this.countOfRepeats = countOfRepeats;
        totalCountOfRepeats = countOfRepeats;
        if (countOfRepeats > 0) {
            nextTrackInfoDisabled = true;
        }
    }

    void clearQueue()                                               { trackQueue.clear(); }
    public int getQueueSize()                                       { return trackQueue.size(); }
    void cancelIdleThread()                                         { threadCountToLeave.cancel(true); }
    void addToQueue(AudioQueueExtendedInfo track)                   { trackQueue.add(track); }
    void setCurrentPausedTrack()                                    { pausedTrack = audioPlayer.getPlayingTrack(); }
    void clearPausedTrack()                                         { pausedTrack = null; }
    void setInfiniteRepeating(boolean infiniteRepeating)            { this.infiniteRepeating = infiniteRepeating; }
    int decreaseCountOfRepeats()                                    { return --countOfRepeats; }
    int getCurrentRepeat()                                          { return (totalCountOfRepeats - countOfRepeats) + 1; }
    void setNextTrackInfoDisabled(boolean nextTrackInfoDisabled)    { this.nextTrackInfoDisabled = nextTrackInfoDisabled; }
}

File: src/main/kotlin/pl/jwizard/core/audio/scheduler/SchedulerActions.kt
/*
 * Copyright (c) 2024 by JWizard
 * Originally developed by Miłosz Gilga <https://miloszgilga.pl>
 */
package pl.jwizard.core.audio.scheduler

import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import net.dv8tion.jda.api.entities.Member
import pl.jwizard.core.audio.ExtendedAudioTrackInfo
import pl.jwizard.core.audio.TrackPosition
import pl.jwizard.core.bot.BotConfiguration
import pl.jwizard.core.command.embed.CustomEmbedBuilder
import pl.jwizard.core.db.GuildDbProperty
import pl.jwizard.core.db.RadioStationDto
import pl.jwizard.core.exception.AudioPlayerException
import pl.jwizard.core.i18n.I18nResLocale
import pl.jwizard.core.log.AbstractLoggingBean
import pl.jwizard.core.util.DateUtils
import pl.jwizard.core.util.Formatter
import java.util.*
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

class SchedulerActions(
	private val botConfiguration: BotConfiguration,
	private val audioScheduler: AudioScheduler,
) : AbstractLoggingBean(SchedulerActions::class) {

	val trackQueue: Queue<AudioTrack> = LinkedList()
	private var totalCountOfRepeats = 0

	var threadsCountToLeave: ScheduledFuture<*>? = null
		private set
	var onClearing: Boolean = false
		private set
	var infiniteRepeating = false
		private set
	var infinitePlaylistRepeating = false
		private set
	var pausedTrack: AudioTrack? = null
	var countOfRepeats = 0
	var nextTrackInfoDisabled = false
	var radioStationDto: RadioStationDto? = null

	internal fun addToQueueAndOffer(audioTrack: AudioTrack) {
		if (!audioScheduler.audioPlayer.startTrack(audioTrack, true)) {
			trackQueue.offer(audioTrack)
		}
	}

	fun nextTrack() {
		val audioTrack = trackQueue.poll()
		if (audioTrack != null) {
			audioScheduler.audioPlayer.startTrack(audioTrack, false)
		}
	}

	fun skipToPosition(position: Int) {
		var audioTrack: AudioTrack? = null
		for (i in 0..position) {
			audioTrack = trackQueue.poll()
		}
		if (audioTrack != null) {
			audioScheduler.audioPlayer.startTrack(audioTrack, false)
		}
	}

	fun moveToPosition(positions: TrackPosition): AudioTrack {
		val copiedTracks = ArrayList(trackQueue)
		val selectedTrack = copiedTracks.removeAt(positions.previous - 1)
		copiedTracks.add(positions.selected - 1, selectedTrack)
		trackQueue.clear()
		trackQueue.addAll(copiedTracks)
		return selectedTrack
	}

	fun removeAllTracksFromMember(member: Member): List<AudioTrack> {
		return ArrayList(trackQueue)
			.filter { (it.userData as Member).id == member.id }
			.mapNotNull { trackQueue.poll() }
	}

	fun checkIfAllTracksIsFromSelectedMember(member: Member): Boolean {
		if (trackQueue.isEmpty()) {
			if (audioScheduler.audioPlayer.playingTrack == null) {
				return false
			}
			val sender = audioScheduler.audioPlayer.playingTrack.userData as Member
			return sender.id == member.id
		}
		return trackQueue.all { (it.userData as Member).id == member.id }
	}

	fun clearAndDestroy(showMessage: Boolean) {
		onClearing = true
		audioScheduler.audioPlayer.isPaused = false
		audioScheduler.audioPlayer.stopTrack()
		trackQueue.clear()

		pausedTrack = null
		countOfRepeats = 0
		totalCountOfRepeats = 0
		infiniteRepeating = false
		nextTrackInfoDisabled = false
		infinitePlaylistRepeating = false

		if (showMessage) {
			val messageEmbed = CustomEmbedBuilder(botConfiguration, audioScheduler.event).buildBaseMessage(
				placeholder = I18nResLocale.LEAVE_EMPTY_CHANNEL
			)
			audioScheduler.event.instantlySendEmbedMessage(messageEmbed, legacyTransport = true)
		}
		onClearing = false
		jdaLog.info(audioScheduler.event, "Remove playing track and clear queue")
	}

	fun leaveAndSendMessageAfterInactivity() {
		val timeToLeaveChannel = botConfiguration.guildSettingsSupplier
			.fetchDbProperty(GuildDbProperty.LEAVE_NO_TRACKS_SEC, audioScheduler.event.guildId, Int::class)
		threadsCountToLeave = botConfiguration.threadPool.schedule({
			val messageEmbed = CustomEmbedBuilder(botConfiguration, audioScheduler.event).buildBaseMessage(
				I18nResLocale.LEAVE_END_PLAYBACK_QUEUE,
				params = mapOf("elapsed" to DateUtils.convertSecToMin(timeToLeaveChannel.toLong()))
			)
			if (!audioScheduler.lockedGuilds.contains(audioScheduler.event.guildId)) {
				clearAndDestroy(false)

				val guild = audioScheduler.event.dataSender?.guild
				botConfiguration.threadPool.submit { guild?.audioManager?.closeAudioConnection() }
				log.info("Audio connection threadpool for guild: {} was closed", Formatter.guildTag(guild))

				audioScheduler.event.instantlySendEmbedMessage(messageEmbed, legacyTransport = true)
				jdaLog.info(
					audioScheduler.event,
					"Leaved voice channel after $timeToLeaveChannel seconds of inactivity"
				)
			}
		}, timeToLeaveChannel.toLong(), TimeUnit.SECONDS)
	}

	fun getAverageTracksDuration(): Long = trackQueue
		.map { it.duration }
		.average()
		.toLong()

	fun getTrackByPosition(position: Int): AudioTrack = ArrayList(trackQueue)[position - 1]

	fun checkInvertedTrackPosition(position: Int) = position <= 0 || position > trackQueue.size

	fun currentRepeat() = (totalCountOfRepeats - countOfRepeats) + 1

	fun checkIfMemberAddAnyTracksToQueue(member: Member) = trackQueue
		.any { (it.userData as Member).id == member.id }

	fun toggleInfiniteRepeating(): Boolean {
		infiniteRepeating = !infiniteRepeating
		return infiniteRepeating
	}

	fun toggleInfinitePlaylistRepeating(): Boolean {
		infinitePlaylistRepeating = !infinitePlaylistRepeating
		return infinitePlaylistRepeating
	}

	fun updateCountOfRepeats(count: Int) {
		countOfRepeats = count
		totalCountOfRepeats = count
		if (count > 0) {
			nextTrackInfoDisabled = true
		}
	}

	fun getPausedTrackInfo(): ExtendedAudioTrackInfo = pausedTrack
		?.let { ExtendedAudioTrackInfo(it) }
		?: throw AudioPlayerException.TrackIsNotPausedException(audioScheduler.event)
}
